llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp

```diff
@@ -4014,6 +4020,50 @@ bool AArch64DAGToDAGISel::SelectSVEAddSubImm(SDValue N, MVT VT, SDValue &Imm,
   return false;
 }
 
+bool AArch64DAGToDAGISel::SelectSVEAddSubSSatImm(SDValue N, MVT VT,
+                                                 SDValue &Imm, SDValue &Shift) {
+  if (!isa<ConstantSDNode>(N))
+    return false;
+
+  SDLoc DL(N);
+  int64_t Val = cast<ConstantSDNode>(N)
+                    ->getAPIntValue()
+                    .trunc(VT.getFixedSizeInBits())
+                    .getSExtValue();
+
+  // Signed saturating instructions treat their immediate operand as unsigned.
+  if (Val < 0)
+    return false;
+
+  switch (VT.SimpleTy) {
```

Comment:
This switch looks the same as the one in `SelectSVEAddSubImm` above, would it be possible to use the same function for the signed instructions and pass an extra `Signed` flag for checking `Val < 0`?

Reply:
I thought the same but whilst the switch is the same, the type of `Val` is different.  For `SelectSVEAddSubImm` it is unsigned but here it is signed.